src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,120 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPASITY = 10;
+    private static final int NOT_FOUND = -1;
+    private static final int GROW_FACTOR = 2;
+    private int size;
+    private T[] array;
+
+    public ArrayList() {
+        array = (T[]) new Object[DEFAULT_CAPASITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == array.length) {
+            grow();
+        }
+        array[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (invalidIndexToAdd(index)) {
+            throw new ArrayListIndexOutOfBoundsException(index + " is out of range index");
+        }
+        grow();
+        for (int i = size; i > index; i--) {
+            array[i] = array[i - 1];
+        }
+        array[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkValidIndexToGetSet(index);
+        return array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkValidIndexToGetSet(index);
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T value = get(index);
+        System.arraycopy(array, index + 1, array, index, size - index - 1);
+        size--;
+        return value;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = find(element);
+        if (index == NOT_FOUND) {
+            throw new NoSuchElementException("Element " + element + " was not found");
+        }
+        return remove(index);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void grow() {
+        if (size < array.length) {
+            return;
+        }
+        int oldLength = array.length;
+        int newLength = oldLength + oldLength / GROW_FACTOR;
+        T[] newArr = (T[]) new Object[newLength];
+        System.arraycopy(array, 0, newArr, 0, oldLength);
+        array = newArr;
+    }
```

Comment:
the logic of your grow method is slightly unclear. If the name is 'grow' then it should always increase the size. Please either change the method name or the inner logic

Reply:
\+